src/game/Movement/WaypointMovementGenerator.cpp

```diff
@@ -181,64 +203,111 @@ bool WaypointMovementGenerator<Creature>::OnArrived(Creature& creature)
     return true;
 }
 
+// minimum time that will take the unit to travel the path (much higher value have been seen in sniff)
+static const uint32 MinimumPathTime = 6000;
+// client need to receive new path before the end of previous path (much higher value have been seen in sniff)
+static const uint32 PreSendTime     = 1500;
+
 void WaypointMovementGenerator<Creature>::StartMove(Creature &creature)
 {
+    m_nodeIndexes.clear(); // make sure to reset spline indexes as its a new path
+
     if (!i_path || i_path->empty())
         return;
 
     if (Stopped())
         return;
 
+    m_creatureSpeed = creature.GetSpeed(Movement::SelectSpeedType(creature.m_movementInfo.GetMovementFlags()));
     WaypointPath::const_iterator currPoint = i_path->find(i_currentNode);
-    MANGOS_ASSERT(currPoint != i_path->end());
 
-    if (m_isArrivalDone)
+    if (currPoint == i_path->end())
     {
-        bool reachedLast = false;
-        ++currPoint;
-        if (currPoint == i_path->end())
-        {
-            reachedLast = true;
-
-            if (!m_repeating)
-            {
-                creature.SetHomePosition(i_path->at(i_currentNode).x, i_path->at(i_currentNode).y, i_path->at(i_currentNode).z, i_path->at(i_currentNode).orientation);
-                creature.GetMotionMaster()->Initialize();
-                return;
-            }
-
-            currPoint = i_path->begin();
-        }
-
-        i_currentNode = currPoint->first;
+        creature.SetHomePosition(i_path->at(m_lastReachedWaypoint).x, i_path->at(m_lastReachedWaypoint).y, i_path->at(m_lastReachedWaypoint).z, i_path->at(m_lastReachedWaypoint).orientation);
+        creature.GetMotionMaster()->Initialize();
+        return;
     }
 
-    m_isArrivalDone = false;
     creature.AddUnitState(UNIT_STATE_ROAMING_MOVE);
 
-    WaypointNode const& nextNode = currPoint->second;
     Movement::MoveSplineInit init(creature, "WaypointMovementGenerator<Creature>::StartMove");
-
-    if (WaypointPath const* pSubpath = (nextNode.path_id ? sWaypointMgr.GetPathFromOrigin(nextNode.path_id, 0, 0, PATH_FROM_SPECIAL) : nullptr))
+    PointsArray genPath; // will contain the generated path
+    genPath.reserve(20); // little optimization
+    genPath.push_back(Vector3(creature.GetPositionX(), creature.GetPositionY(), creature.GetPositionZ()));
```

Comment:
Consider using emplace_back when applicable. Here it avoids an unnecessary copy.
```suggestion
    genPath.emplace_back(creature.GetPositionX(), creature.GetPositionY(), creature.GetPositionZ());
```